ruby/ql/lib/codeql/ruby/frameworks/ActiveStorage.qll

```diff
@@ -8,23 +8,236 @@ private import codeql.ruby.Concepts
+   * ```rb
+   * class User < ActiveRecord::Base
+   *   has_one_attached :avatar
+   * end
+   *
+   * user = User.find(id)
+   * user.avatar
+   * ```
+   */
+  private class AttachmentInstance extends DataFlow::CallNode {
+    AttachmentInstance() {
+      exists(Association assoc, string model | model = assoc.getTargetModelName() |
+        this.getReceiver().(ActiveRecordInstance).getClass() = assoc.getSourceClass() and
+        (
+          assoc.isSingular() and this.getMethodName() = model
+          or
+          assoc.isCollection() and this.getMethodName() = model
+        )
+      )
+    }
+  }
+
+  /**
+   * A call on an ActiveStorage  object that results in an image transformation.
+   * Arguments to these calls may be executed as system commands.
+   */
+  private class ImageProcessingCall extends DataFlow::CallNode, SystemCommandExecution::Range {
+    ImageProcessingCall() {
+      this =
+        ModelOutput::getATypeNode("activestorage", "Blob")
+            .getAMethodCall(["variant", "preview", "representation"]) or
+      this =
+        API::getTopLevelMember("ActiveStorage")
+            .getMember("Attachment")
+            .getInstance()
+            .getAMethodCall(["variant", "preview", "representation"]) or
+      this =
+        API::getTopLevelMember("ActiveStorage")
+            .getMember("Variation")
+            .getAMethodCall(["new", "wrap", "encode"]) or
+      this =
+        API::getTopLevelMember("ActiveStorage")
+            .getMember("Variation")
+            .getInstance()
+            .getAMethodCall("transformations=") or
+      this =
+        API::getTopLevelMember("ActiveStorage")
+            .getMember("Transformers")
+            .getMember("ImageProcessingTransformer")
+            .getAMethodCall("new") or
+      this =
+        API::getTopLevelMember("ActiveStorage")
+            .getMember(["Preview", "VariantWithRecord"])
+            .getAMethodCall("new") or
+      // `ActiveStorage.paths` is a global hash whose values are passed to
+      // a `system` call.
+      this = API::getTopLevelMember("ActiveStorage").getAMethodCall("paths=") or
+      // `ActiveStorage.video_preview_arguments` is passed to a `system` call.
+      this = API::getTopLevelMember("ActiveStorage").getAMethodCall("video_preview_arguments=")
+    }
+
+    override DataFlow::Node getAnArgument() { result = this.getArgument(0) }
+  }
+
+  /**
+   * `ActiveStorage.variant_processor` is passed to `const_get`.
+   */
+  private class VariantProcessor extends DataFlow::CallNode, CodeExecution::Range {
+    VariantProcessor() {
+      this = API::getTopLevelMember("ActiveStorage").getAMethodCall("variant_processor=")
+    }
+
+    override DataFlow::Node getCode() { result = this.getArgument(0) }
+  }
+
+  /**
+   * Adds ActiveStorage instances to the API graph.
+   * Source code may not mention `ActiveStorage` or `ActiveStorage::Attachment`,
+   * so we add synthetic nodes for them.
+   */
+  private module ApiNodes {
```

Comment:
Before MaD, we would generally handle this by introducing a predicate returning the set of API nodes of a certain type. For example, we'd have:
```ql
API::Node activeStorageStatic() {
  result = API::getTopLevelMember("ActiveStorage")
}

API::Node attachmentStatic() {
  result = activeStorageStatic().getMember("Attachment")
}

API::Node attachmentInstance() {
  result = attachmentStatic().getInstance()
  or
  result = any(AttachmentInstanceEntryPoint e).getANode()
}

class AttachmentInstanceEntryPoint extends API::EntryPoint {
  AttachmentInstanceEntryPoint() { this = "AttachmentInstanceEntryPoint" }

  DataFlow::LocalSourceNode getAUse() { result instanceof AttachmentInstance }
}
```

And then instead of writing out `API::getTopLevelMember("ActiveStorage").getMember("Attachment").getInstance()` we'd use a call to `attachmentInstance()`. (With MaD, this all becomes type-definitions rows instead, but there is no pressure to use MaD right now)

Is there a reason to add this new way of doing it? One thing I worry about is that the choice of access path is arbitrary, and there's no way to know which path you want to use at the use-site. For example, suppose `ABC::Attachment` was an alias for `ActiveStorage::Attachment`, would you expect `API::getTopLevelMember("ABC").getMember("Attachment")` to work as well?

Reply:
After thinking about this some more, I realised that the API graph edge relation this now mutually recursive with things like `getTopLevelMember` and `getMember` and each of these `EntryPoint.edge` predicates.

For instance, this means this implementation,
```codeql
      override predicate edge(API::Node pred, API::Label::ApiLabel lbl) {
        pred = API::getTopLevelMember("ActiveStorage").getMember("Attachment").getMethod("new") and
        lbl = API::Label::return()
      }
```
is a recursive predicate with a three-way recursive join (2x `getMember` and 1x `getTopLevelMember`). This is pretty risky performance-wise, and even if it appears to scale now it's the kind of thing that can become a problem later.

Reply:
Thanks! I'm [working on adding](https://github.com/github/codeql/pull/10276) the missing hook so you can contribute to the MaD type-def from CodeQL. I expect it would look something like this,
```codeql
class AttachmentInstanceType extends ModelInput::TypeModel {
  override DataFlow::LocalSourceNode getASource(string package, string type) {
    package = "activestorage" and
    type = "Attachment" and
    result instanceof AttachmentInstance
  }
}
```
Then you could write the MaD lines like this
```js
// ActiveStorage::Attachment#blob : Blob    
"activestorage;Blob;activestorage;Attachment;Method[blob].ReturnValue",
// ActiveStorage::Attachment delegates method calls to its associated Blob       
"activestorage;Blob;activestorage;Attachment;",
```

But if you can work around it for now that would be great.

Reply:
[The PR with TypeModel](https://github.com/github/codeql/pull/10276) has been merged, it would be great if you could use it in this PR, but I'll leave it up to you.